Create a new 'live' ZAP docker image based on the latest code in the repos

psiinon opened the pull request:
Will upload a version of this image to https://hub.docker.com/r/psiinon/ for people to try out once its finished building (and will link from this issue).
The plan is to auto generate this image whenever the zaproxy repo is changed as per https://docs.docker.com/docker-hub/builds/
I'm also hoping we can use it to auto generate daily builds using https://wiki.mozilla.org/TaskCluster

thc202 commented on build/docker/Dockerfile-develop:
```diff
@@ -0,0 +1,58 @@
+# This dockerfile allows you to build and run OWASP ZAP using the latest source code
+# To build use a command line:
+#	docker build -t owasp/zap2docker-dev -f Dockerfile-develop .
+# To use use a command like
```
`To run use`?

thc202 commented on build/docker/Dockerfile-develop:
```diff
@@ -0,0 +1,58 @@
+# /zap/src	will contain the ZAP repos with the latest source code (from when the container was built)
+# /zap/zap will contain a version of ZAP built from the source code
+#
+# If you change any of the ZAP extensions the you can rebuild them using:
```
`then you can`?

thc202 commented on build/docker/Dockerfile-develop:
```diff
@@ -0,0 +1,58 @@
+  && ln -s /usr/share/ant/bin/ant /usr/bin/ant
+ENV ANT_HOME /usr/share/ant
+
+# Install vi so that its easy to make quick changes 
```
`Install vim`

thc202 commented on build/docker/Dockerfile-develop:
```diff
@@ -0,0 +1,58 @@
+
+# Pull the ZAP repos
+RUN git clone https://github.com/zaproxy/zaproxy.git && \
+	git clone https://github.com/zapbot/zap-extensions.git && \
```
Should be from zaproxy? (same comment for following clones)

psiinon commented:
Good spots :) Hopefully fixed now

jvehent commented on build/docker/Dockerfile-develop:
```diff
@@ -0,0 +1,52 @@
+RUN	ln -s /zap/src/zaproxy/build/zap /zap
+
+# Sanity check that we can run ZAP	
+RUN cd /zap/zap && ./zap.sh -version
```
Every dockerfile instruction creates a new layer inside the container, so you can significantly reduce the container size by merging those into a single statement. For example: https://github.com/mozilla-services/hindsight/blob/master/Dockerfile#L20-L62

It might also be useful to delete the source code after the build and remove dev packages to help reduce the image size.

psiinon replied:
The idea of this image was to allow people to start hacking on the ZAP source code easily, so deleting the source wouldnt be good ;)
However ... we could just have another image which was based on this and just deleted the source. And, thinking about it more, this should include all of the other sw that the weekly release includes (if its to replace the weekly release).
So the image should really be much closer to the weekly one :/
I'll do some more playing ...

jvehent replied:
Ha yes, you'd want to keep the build tools then.

I quite like the idea of a lean container for testers and a larger one for developers. I think you could use the same dockerfile for both and alter the build using env variables: http://stackoverflow.com/questions/19537645/get-environment-variable-value-in-dockerfile#34600106

psiinon commented:
Here the new live one: https://hub.docker.com/r/psiinon/zap2docker-live/

psiinon commented:
Changed the PR to just include a new 'live' image - this one does not include the source so I've changed the title.
Example available at https://hub.docker.com/r/psiinon/zap2docker-live/
The plan is to use dockerhubs automated build to build this every day or on every change. We can keep maintaining the weekly by duplicating the Monday morning build (assuming no one has reported any issues).
Ready to review now :)
Future changes may include pushing a daily (non docker) release to github, but I've not looked into that yet.
